From 8ccff99edeb2af6967945ef5ae473b605012af88 Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Tue, 21 Jul 2020 13:19:28 +0800
Subject: [PATCH] PONRTSYS-7743:XPCS: Dynamic detection each linkup event

Link partner speed can change dynamically and need
to reconfigure serdes and switch to link partner
speed,XPCS driver poll the 10G and SGMII link status,
If 10G and SGMII link is down switch to higest speed,
and configure serdes accordingly.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 drivers/net/ethernet/lantiq/xpcs/xpcs.c       | 71 ++++++++++++++++++++-------
 drivers/net/ethernet/lantiq/xpcs/xpcs.h       | 33 ++-----------
 drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c |  5 ++
 3 files changed, 62 insertions(+), 47 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.c b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
index d26e32724294..8d587e40e218 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.c
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.c
@@ -1589,13 +1589,51 @@ static int xpcs_ethtool_ksettings_set(struct device *dev,
 	return 0;
 }
 
+static bool get_sgmii_link_status(struct xpcs_prv_data *pdata)
+{
+	int ret;
+
+	ret = XPCS_RGRD(pdata, SR_MII_STS);
+	if (ret < 0)
+		return ret;
+	/* Link status is latched low, so read again to get
+	 * current state
+	 */
+	ret = XPCS_RGRD(pdata, SR_MII_STS);
+	if (ret < 0)
+		return ret;
+
+	if (!(ret & XPCS_SGMII_LINK_STS))
+		return false;
+
+	return true;
+}
+static bool get_teng_link_status(struct xpcs_prv_data *pdata)
+{
+	int ret;
+
+	ret = XPCS_RGRD(pdata, PCS_KR_STS1);
+	if (ret < 0)
+		return ret;
+	/* Link status is latched low, so read again to get
+	 * current state
+	 */
+	ret = XPCS_RGRD(pdata, PCS_KR_STS1);
+	if (ret < 0)
+		return ret;
+
+	if (!(ret & XPCS_KR_LINK_STS))
+		return false;
+
+	return true;
+}
 /* Poll the link state and reconfigure serdes */
 static void xpcs_link_poll_work(struct work_struct *work)
 {
 	struct xpcs_prv_data *pdata = container_of(work, struct xpcs_prv_data,
 						link_change_work.work);
-	bool link_status = false;
 	u32 mode = 0;
+	bool teng_link = true, sgmii_link = true;
 
 	/* Since the device can be in the wrong mode when a link is
 	 * (re-)established (SFP connected after the interface is
@@ -1607,27 +1645,22 @@ static void xpcs_link_poll_work(struct work_struct *work)
 		pdata->link_mode = 0;
 
 	mode = pdata->link_change_seq[pdata->link_mode];
-	xpcs_reinit(pdata->dev, mode);
 	pdata->link_mode++;
 
-	switch (mode) {
-	case TENG_KR_MODE:
-		link_status  = get_xpcs_link_status(pdata) > 0 ? 1 : 0;
-		break;
-	case TENG_XAUI_MODE:
-		break;
-	case ONEG_XAUI_MODE:
-	case TWOP5G_SGMII_MODE:
-		link_status = get_sgmii_link_status(pdata) > 0 ? 1 : 0;
-		break;
-	}
+	if (!get_teng_link_status(pdata))
+		teng_link = false;
+	if (!get_sgmii_link_status(pdata))
+		sgmii_link = false;
 
-	if (!link_status) {
-		/* Reschedule work for every poll interval */
-		queue_delayed_work(pdata->link_poll,
-				   &pdata->link_change_work,
-				   (HZ * pdata->poll_interval));
-	}
+	if (!(teng_link || sgmii_link) && mode != TENG_XAUI_MODE)
+		xpcs_reinit(pdata->dev, mode);
+	else
+		pdata->link_mode = 0;
+
+	/* Reschedule work for every poll interval */
+	queue_delayed_work(pdata->link_poll,
+			   &pdata->link_change_work,
+			   (HZ * pdata->poll_interval));
 }
 
 static int dynamic_link_detection(struct xpcs_prv_data *pdata)
diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs.h b/drivers/net/ethernet/lantiq/xpcs/xpcs.h
index b3486ca6a895..2d4cd791bab0 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs.h
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs.h
@@ -567,8 +567,11 @@
 #define PMA_RX_EQ_CTRL0_VGA1_GAIN_0_POS	12
 #define PMA_RX_EQ_CTRL0_VGA1_GAIN_0_WIDTH	4
 
-#define PCS_KR_STS1_PLU_POS	12
-#define PCS_KR_STS1_PLU_WIDTH	1
+#define PCS_KR_STS1_RPCS_BKLK_POS    0
+#define PCS_KR_STS1_RPCS_BKLK_WIDTH  1
+
+#define XPCS_KR_LINK_STS 0x0001
+#define XPCS_SGMII_LINK_STS 0x0004
 
 enum {
 	BACKPL_ETH_PCS = 0,
@@ -929,31 +932,5 @@ static inline struct xpcs_ops *xpcs_pdev_ops(struct platform_device *pdev)
 	return &pdata->ops;
 }
 
-static inline int get_sgmii_link_status(struct xpcs_prv_data *pdata)
-{
-	bool link_status = false;
-
-	link_status = XPCS_RGRD_VAL(pdata, SR_MII_STS, LINK_STS);
-	/* Link status is latched low, so read again to get
-	 * current state
-	 */
-	link_status = XPCS_RGRD_VAL(pdata, SR_MII_STS, LINK_STS);
-
-	return link_status;
-}
-
-static inline int get_xpcs_link_status(struct xpcs_prv_data *pdata)
-{
-	bool link_status = false;
-
-	link_status = XPCS_RGRD_VAL(pdata, PCS_KR_STS1, PLU);
-	/* Link status is latched low, so read again to get
-	 * current state
-	 */
-	link_status = XPCS_RGRD_VAL(pdata, PCS_KR_STS1, PLU);
-
-	return link_status;
-}
-
 #endif
 
diff --git a/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c b/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c
index 9990affabf3a..b07854a3b5be 100644
--- a/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c
+++ b/drivers/net/ethernet/lantiq/xpcs/xpcs_sysfs.c
@@ -237,6 +237,11 @@ xpcs_linksts_show(struct device *dev,
 	val = XPCS_RGRD_VAL(pdata, VR_MII_AN_CTRL, SGMII_LINK_STS);
 	off += sprintf(buf + off, "\tSGMII/QSGMII/USXGMII mode Port0 Rx:  %s\n",
 		       val ? "LINK UP" : "LINK DOWN");
+	off += sprintf(buf + off, "\n10GBASE-R Status\n");
+	off += sprintf(buf + off, "\n");
+	val = XPCS_RGRD_VAL(pdata, PCS_KR_STS1, RPCS_BKLK);
+	off += sprintf(buf + off, "\t10GBASE-R:  %s\n",
+		       val ? "LINK UP" : "LINK DOWN");
 
 	return off;
 }
